subsys/net/lib/dhcp_server/dhcpv4_server.c

```diff
@@ -0,0 +1,1423 @@
+	int ret;
+
+	if (datalen < sizeof(struct dhcp_msg)) {
+		LOG_DBG("DHCPv4 server malformed message");
+		return;
+	}
+
+	msg = (struct dhcp_msg *)data;
+
+	if (msg->op != DHCPV4_MSG_BOOT_REQUEST) {
+		/* Silently drop messages other than BOOTREQUEST */
+		return;
+	}
+
+	data += sizeof(struct dhcp_msg);
+	datalen -= sizeof(struct dhcp_msg);
+
+	/* Skip server hostname/filename/option cookie */
+	if (datalen < (SIZE_OF_SNAME + SIZE_OF_FILE + SIZE_OF_MAGIC_COOKIE)) {
+		return;
+	}
+
+	data += SIZE_OF_SNAME + SIZE_OF_FILE + SIZE_OF_MAGIC_COOKIE;
+	datalen -= SIZE_OF_SNAME + SIZE_OF_FILE + SIZE_OF_MAGIC_COOKIE;
+
+	/* Search options for DHCP message type. */
+	ret = dhcpv4_find_message_type_option(data, datalen, &msgtype);
+	if (ret < 0) {
+		LOG_ERR("No message type option");
+		return;
+	}
+
+	k_mutex_lock(&server_lock, K_FOREVER);
+
+	switch (msgtype) {
+	case NET_DHCPV4_MSG_TYPE_DISCOVER:
+		dhcpv4_handle_discover(ctx, msg, data, datalen);
+		break;
+	case NET_DHCPV4_MSG_TYPE_REQUEST:
+		dhcpv4_handle_request(ctx, msg, data, datalen);
+		break;
+	case NET_DHCPV4_MSG_TYPE_DECLINE:
+		dhcpv4_handle_decline(ctx, msg, data, datalen);
+		break;
+	case NET_DHCPV4_MSG_TYPE_RELEASE:
+		dhcpv4_handle_release(ctx, msg, data, datalen);
+		break;
+	case NET_DHCPV4_MSG_TYPE_INFORM:
+		dhcpv4_handle_inform(ctx, msg, data, datalen);
+		break;
+
+	case NET_DHCPV4_MSG_TYPE_OFFER:
+	case NET_DHCPV4_MSG_TYPE_ACK:
+	case NET_DHCPV4_MSG_TYPE_NAK:
+	default:
+		/* Ignore server initiated and unknown message types. */
+		break;
+	}
+
+	k_mutex_unlock(&server_lock);
+}
+
+static void dhcpv4_server_cb(struct k_work *work)
+{
+	struct net_socket_service_event *evt =
+		CONTAINER_OF(work, struct net_socket_service_event, work);
+	struct dhcpv4_server_ctx *ctx = NULL;
+	uint8_t recv_buf[NET_IPV4_MTU];
+	int ret;
+
+	for (int i = 0; i < ARRAY_SIZE(server_ctx); i++) {
+		if (server_ctx[i].sock == evt->event.fd) {
+			ctx = &server_ctx[i];
+			break;
+		}
+	}
+
+	if (ctx == NULL) {
+		LOG_ERR("No DHCPv4 server context found for given FD.");
+		return;
+	}
+
+	if (evt->event.revents & ZSOCK_POLLERR) {
+		LOG_ERR("DHCPv4 server poll revents error");
+		net_dhcpv4_server_stop(ctx->iface);
+		return;
+	}
+
+	if (!(evt->event.revents & ZSOCK_POLLIN)) {
+		return;
+	}
+
+	ret = zsock_recvfrom(evt->event.fd, recv_buf, sizeof(recv_buf),
+			     ZSOCK_MSG_DONTWAIT, NULL, 0);
+	if (ret < 0) {
+		if (errno == EAGAIN) {
+			return;
+		}
+
+		LOG_ERR("DHCPv4 server recv error, %d", errno);
+		net_dhcpv4_server_stop(ctx->iface);
+		return;
+	}
+
+	dhcpv4_process_data(ctx, recv_buf, ret);
+}
+
+NET_SOCKET_SERVICE_SYNC_DEFINE_STATIC(dhcpv4_server, NULL, dhcpv4_server_cb,
+				      CONFIG_NET_DHCPV4_SERVER_INSTANCES);
+
+int net_dhcpv4_server_start(struct net_if *iface, struct in_addr *base_addr)
+{
+	struct sockaddr_in addr = {
+		.sin_family = AF_INET,
+		.sin_addr = INADDR_ANY_INIT,
+		.sin_port = htons(DHCPV4_SERVER_PORT),
+	};
+	struct ifreq ifreq = { 0 };
+	bool start = true;
```

Comment:
Variable doesn't seem to be used?

Reply:
Yes, a leftover from before I've switched to use socket services, thanks.